File: TERCERO/PREPARACION PARA EL FINAL/Febrero2022/java/src/exercises/EdRank.java
/* ------------------------------------------------------------------------------
   -- Student's name:
   -- Student's group:
   -- Identity number (DNI if Spanish/passport if Erasmus):
   --
   -- Data Structures. Grado en Informática. UMA.
   -------------------------------------------------------------------------------
*/

package exercises;

import dataStructures.dictionary.AVLDictionary;
import dataStructures.dictionary.Dictionary;
import dataStructures.graph.DiGraph;
import dataStructures.set.Set;

public class EdRank<T extends Comparable<? super T>> {
    private static final double THRESHOLD = 0.0001;
    private DiGraph<T> diGraph;
    private Dictionary<T, Double> rankDict;

    public EdRank(DiGraph<T> g) {
        diGraph=g;
        rankDict=new AVLDictionary<>();
        for(T v: diGraph.vertices()){
            rankDict.insert(v,0.0);
        }
    }

    public void edRank(double rank) {
        //TODO
        for(T v: diGraph.vertices()) {
            distribute(v, rank);
        }
    }

    private void distribute(T node, double rank) {
        //TODO
        if(rank>THRESHOLD) {
            double rankby2 = rank/2.0;
            rankDict.insert(node, rankDict.valueOf(node)+rankby2);
            double sumageneral = rankby2 / diGraph.successors(node).size();
            for(T v : diGraph.successors(node)) {
                distribute(v,  sumageneral);
            }
        }
    }

    public Dictionary<T, Double> edRank() {
        return rankDict;
    }
}

File: TERCERO/PREPARACION PARA EL FINAL/Febrero2022/haskell/EdRank.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
-- Identity number (DNI if Spanish/passport if Erasmus):
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module EdRank where

import DataStructures.Graph.DiGraph 
import DataStructures.Dictionary.AVLDictionary as D
import Data.Maybe(fromJust)

threshold = 0.0001
createRankDict :: (Ord a) => DiGraph a -> D.Dictionary a Double
createRankDict node = foldr (`D.insert` 0) D.empty (vertices node)

distribute :: Ord a => DiGraph a -> a -> Double -> D.Dictionary a Double -> D.Dictionary a Double
distribute g v rank dict 
    |rank>threshold= updateValues g rankbysuc (successors g v) (D.insert v (rankby2+(fromJust (D.valueOf v dict))) dict)
    |otherwise = dict
        where
            rankby2 = rank*0.5
            rankbysuc = rankby2/ (fromIntegral(length (successors g v)))

updateValues :: Ord a => DiGraph a -> Double -> [a] -> D.Dictionary a Double -> D.Dictionary a Double
updateValues g rk [] dict = dict
updateValues g rk (x:xs) dict = distribute g x rk (updateValues g rk xs dict)

edRank :: Ord a => DiGraph a -> Double -> D.Dictionary a Double
edRank g rank = updateValues g rank (vertices g) (createRankDict g)